pkg/util/image/status_to_label.go

```diff
@@ -4,54 +4,66 @@
 package image
 
 import (
+	"reflect"
+	"strings"
+
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/validation"
 
 	vmopv1 "github.com/vmware-tanzu/vm-operator/api/v1alpha3"
 )
 
-// SyncStatusToLabels copies the image's capabilities and OS information from
-// the image status to well-known labels on the image in order to make it
+// SyncStatusToLabels copies the image's type, capabilities and OS information
+// from the image status to well-known labels on the image in order to make it
 // easier for clients to search for images based on capabilities or OS details.
+// If the image status does not have the info or is invalid, remove the label.
 func SyncStatusToLabels(
 	obj metav1.Object,
 	status vmopv1.VirtualMachineImageStatus) {
 
-	var (
-		osID      = status.OSInfo.ID
-		osType    = status.OSInfo.Type
-		osVersion = status.OSInfo.Version
-		caps      = status.Capabilities
-	)
-
-	if osID == "" && osType == "" && osVersion == "" && len(caps) == 0 {
-		return
-	}
-
 	labels := obj.GetLabels()
 	if labels == nil {
 		labels = map[string]string{}
 	}
 
-	if k, v := vmopv1.VirtualMachineImageOSIDLabel, osID; isValidLabelValue(v) {
-		labels[k] = v
-	}
-	if k, v := vmopv1.VirtualMachineImageOSTypeLabel, osType; isValidLabelValue(v) {
-		labels[k] = v
+	setLabel := func(k, v string) {
+		if isValidLabelValue(v) {
+			labels[k] = v
+		} else {
+			delete(labels, k)
+		}
 	}
-	if k, v := vmopv1.VirtualMachineImageOSVersionLabel, osVersion; isValidLabelValue(v) {
-		labels[k] = v
+
+	// Set image type label.
+	setLabel(vmopv1.VirtualMachineImageTypeLabel, status.Type)
+
+	// Set OS labels.
+	setLabel(vmopv1.VirtualMachineImageOSIDLabel, status.OSInfo.ID)
+	setLabel(vmopv1.VirtualMachineImageOSTypeLabel, status.OSInfo.Type)
+	setLabel(vmopv1.VirtualMachineImageOSVersionLabel, status.OSInfo.Version)
+
+	// Set capability labels.
+	kp := vmopv1.VirtualMachineImageCapabilityLabel
+	capKeys := map[string]struct{}{}
+	for _, cap := range status.Capabilities {
+		k := kp + cap
+		if len(validation.IsQualifiedName(k)) == 0 {
+			labels[k] = "true"
+			capKeys[k] = struct{}{}
+		} else {
+			delete(labels, k)
+		}
 	}
-	if kp, v := vmopv1.VirtualMachineImageCapabilityLabel, caps; len(v) > 0 {
-		for i := range v {
-			k := kp + v[i]
-			if len(validation.IsQualifiedName(k)) == 0 {
-				labels[k] = "true"
-			}
+	// Delete any stale capability labels.
+	for k := range labels {
+		if _, ok := capKeys[k]; !ok && strings.HasPrefix(k, kp) {
+			delete(labels, k)
 		}
 	}
 
-	obj.SetLabels(labels)
+	if !reflect.DeepEqual(labels, obj.GetLabels()) {
```

Comment:
Why do this instead updating the objects labels in-place (which will already happen when object's labels != nil)

Reply:
Good point. Updated to invoke `SetLabels` if the labels == nil and removed this block at the end. See: https://github.com/vmware-tanzu/vm-operator/compare/0155a2f9769e4842838101f19c2dac8ce6a8a85e..196d3999562481d1893eb4996acca22fadcebcde